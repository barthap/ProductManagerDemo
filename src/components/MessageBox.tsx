import React from "react";
import {StyleSheet, Text, View} from "react-native";
import {Icon} from "native-base";
import {useDispatch } from "react-redux";
import {IMessageBoxState} from "../core/reducers/MessageBoxReducer";
import {typedUseSelector} from "../utils/typedUseSelector";
import {messageBoxActions} from "../core/actions/MessageBox.actions";

interface MessageBoxComponentProps {
    iconShown?: boolean;
    iconName?: typeof Icon.name
    dismissButtonShown?: boolean;
    backgroundColor?: string;
    textColor?: string;
    children: React.ReactNode
    onDismiss?: () => void;
}

//raw "dumb" component
export function MessageBoxComponent(props: MessageBoxComponentProps) {
    const messageStyle = props.textColor ? {...styles.message, color: props.textColor} : styles.message;

    return (<View style={{...styles.container, backgroundColor: props.backgroundColor}}>
        {props.iconShown && <Icon name={props.iconName} style={styles.iconL} fontSize={15} />}
        <Text style={messageStyle}>{props.children}</Text>
        {props.dismissButtonShown && <Icon name="ios-close"
              fontSize={20} style={styles.iconR}
              onPress={props.onDismiss} />}
    </View>);
}

export type MessageBoxStyle = Partial<MessageBoxComponentProps>;
export type AlertStyle = 'info' | 'error' | 'success';

const infoMessageBoxStyle: MessageBoxStyle = {
    iconShown: true,
    iconName: 'ios-information-circle',
    backgroundColor: '#95D4FF',
    textColor: '#024aff',
};
const successMessageBoxStyle: MessageBoxStyle = {
    iconShown: false,
    backgroundColor: '#8dff71',
    textColor: '#00aa01',
};
const errorMessageBoxStyle: MessageBoxStyle = {
    iconShown: true,
    iconName: 'ios-alert',
    backgroundColor: '#ff8273',
    textColor: '#aa0300',
};

function getMessageBoxStyle(styleName: AlertStyle): MessageBoxStyle {
    switch (styleName) {
        case "error":
            return  errorMessageBoxStyle;
        case "info":
            return  infoMessageBoxStyle;
        case "success":
            return  successMessageBoxStyle;
    }
}

export function MessageBox() {
    const mb = typedUseSelector<IMessageBoxState>(state => state.messageBox);
    const style = getMessageBoxStyle(mb.alertStyle);
    const dispatch = useDispatch();

    if(!mb.visible) return <View/>; //When hidden, render empty View
    return (
        <MessageBoxComponent onDismiss={() => dispatch(messageBoxActions.dismiss())}
                             {...style}
                             dismissButtonShown={mb.dismissible}>
            {mb.message}
        </MessageBoxComponent>
    );
}

const styles = StyleSheet.create({
    container: {
        bottom: 0,
        position: 'absolute',
        width: '100%',
        padding: 15,

        flex: 1,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center'
    },
    message: {
        flex: 8,
        fontSize: 15,
        fontWeight: 'bold'
    },
    iconL: {
        flex: 1,
        paddingRight: 5
    },
    iconR: {
        alignSelf: 'flex-end',
        flex: 1,
        paddingLeft: 5
    }
});
