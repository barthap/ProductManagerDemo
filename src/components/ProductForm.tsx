import {Product} from "../api/ProductApi";
import React, {useState} from "react";
import {Form, Item, Label, Input, Button, Text, Textarea, Picker, Icon} from 'native-base';
import {View} from "react-native";
import i18n from "../i18n";

export interface ProductFormProps {
    initialData?: Product;
    onSubmit: (product: Product) => void;
}

//Form has simple example of field validation (for name)
//but it can be done cleaner and more nicely
export function ProductForm(props: ProductFormProps) {

    const initialData = props.initialData || {name: '', id: '-1', quantity: 1, quantityUnit: 'szt', description: ''};
    const [name, setName] = useState(initialData.name || '');
    const [quantity, setQuantity] = useState(initialData.quantity || 0);
    const [quantityUnit, setQuantityUnit] = useState(initialData.quantityUnit || 'szt');
    const [description, setDescription] = useState(initialData.description || '');
    const [nameErr, setNameErr] = useState(false);
    const [errText, setErrText] = useState('');
    const handleSubmit = () => {
        if(name == null || name.length == 0) {
            console.log('empty name');
            setNameErr(true);
            setErrText(i18n.t('form.messages.name_empty'));
            return;
        }
        const id = initialData.id || null;
        props.onSubmit({id, name, quantityUnit, quantity, description})
    };


    return (<React.Fragment>
        <Text style={{color: '#f00'}}>{errText}</Text>
        <Form style={{paddingBottom: 20}}>
            <Item error={nameErr} stackedLabel>
                <Label>{i18n.t('form.fields.name')}</Label>
                <Input value={name} onChangeText={val => {
                    setName(val);
                    if(val.length > 0) {    //reset error when typing
                        setErrText('');
                        setNameErr(false);
                    }
                }}/>
            </Item>
            <View style={{ flexDirection: 'row', flex: 1, alignItems: 'stretch'}}>
                <Item stackedLabel style={{flex: 1}}>
                    <Label>{i18n.t('form.fields.quantity')}</Label>
                    <Input keyboardType="numeric" value={quantity.toString()} onChangeText={val => {
                        const num = parseInt(val);
                        setQuantity(isNaN(num) ? 0 : num);
                    }}/>
                </Item>
                <Item picker>
                    <Picker
                        mode="dropdown"
                        iosIcon={<Icon name="arrow-down" />}
                        style={{ width: undefined }}
                        placeholder={i18n.t('form.fields.unit')}
                        placeholderStyle={{ color: "#bfc6ea" }}
                        placeholderIconColor="#007aff"
                        selectedValue={quantityUnit}
                        onValueChange={setQuantityUnit}
                    >
                        <Picker.Item label={i18n.t('form.units.pieces')} value="szt" />
                        <Picker.Item label={i18n.t('form.units.litres')} value="L" />
                        <Picker.Item label={i18n.t('form.units.kilograms')} value="kg" />
                    </Picker>
                </Item>
            </View>
            <Textarea rowSpan={5} bordered underline={true}
                      placeholder={i18n.t('form.fields.description')}
                      value={description}
                      onChangeText={t => setDescription(t)}/>
        </Form>
        <Button primary onPress={handleSubmit}><Text>{i18n.t('form.save')}</Text></Button>
    </React.Fragment>);
}
