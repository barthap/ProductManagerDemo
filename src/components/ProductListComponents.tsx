import React from 'react';
import {FlatList, StyleSheet} from "react-native";
import {Product} from "../api/ProductApi";
import {ListItem, Text} from 'native-base';

type ProductListProps = {
    items: Array<Product>;
    onItemClick: (product: Product) => void;
}
type ListItemProps = {
    product: Product;
    onClick: (product: Product) => void;
}

function ProductListItem(props: ListItemProps) {
    return (
        <ListItem onPress={() => props.onClick(props.product)}>
            <Text>{props.product.name}</Text>
        </ListItem>
    );
}

export function ProductList(props: ProductListProps) {
    return <FlatList data={props.items}
                     keyExtractor={(item) => item.id.toString() }
                     renderItem={(item) =>
                         <ProductListItem product={item.item} onClick={props.onItemClick}/>
                     } />
}

const styles = StyleSheet.create({

});
