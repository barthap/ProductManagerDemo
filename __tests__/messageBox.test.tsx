import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { MessageBoxComponent } from '../src/components/MessageBox';

jest.mock('@codler/react-native-keyboard-aware-scroll-view');

it('should render MessageBoxComponent', function () {
  const render = renderer.create(<MessageBoxComponent>Hello</MessageBoxComponent>);
  expect(render).toMatchSnapshot();
});

it('should render MessageBoxComponent with icon', function () {
  const render = renderer.create(
    <MessageBoxComponent iconShown iconName="ios-alert">
      Hello
    </MessageBoxComponent>
  );

  expect(render).toMatchSnapshot();
});
