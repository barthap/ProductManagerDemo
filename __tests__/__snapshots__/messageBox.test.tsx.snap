// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MessageBoxComponent 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "bottom": 0,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "padding": 15,
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "flex": 8,
        "fontSize": 15,
        "fontWeight": "bold",
      }
    }
  >
    Hello
  </Text>
</View>
`;

exports[`should render MessageBoxComponent with icon 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "bottom": 0,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "padding": 15,
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <Text />
  <Text
    style={
      Object {
        "flex": 8,
        "fontSize": 15,
        "fontWeight": "bold",
      }
    }
  >
    Hello
  </Text>
</View>
`;
